lenskit-slim/src/main/java/org/lenskit/slim/AbstractLinearRegression.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2016 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.slim;
+
+import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
+import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap;
+import org.grouplens.grapht.annotation.DefaultImplementation;
+import org.lenskit.util.math.Vectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.Map;
+
+import static org.lenskit.slim.LinearRegressionHelper.addVectors;
+
+
+/**
+ * Linear regression which minimize following loss function
+ * 1/2*||a_j - A*w_j||^2 + beta/2*||wj||^2 + lambda*||w_j||
+ */
+@LinearRegression
+@DefaultImplementation(CovarianceUpdateCoordDestLinearRegression.class)
+public abstract class AbstractLinearRegression {
```

Comment:
Since this is - right now - SLIM-specific, what if we called it something SLIM-specific? We'll generalize later.

Maybe call it `SLIMScoringStrategy`?